test/CoreTest.cpp

```diff
@@ -503,11 +519,13 @@ TEST_CASE( "wrapper_test", "[core]" )
     mrc2.loadChunk({0}, {1}, loadData);
     o.flush();
     REQUIRE(loadData[0] == value);
-    value = 43;
+    value = 43.;
     mrc2.makeConstant(value);
-    o.iterations[4].meshes["E"]["y"].loadChunk({0}, {1}, loadData);
+    std::array< double, 1 > moreData = {{ 112233. }};
+    auto pppp = loadRaw(moreData);
```

Comment:
note: we need an `lvalue` since `loadChunk` needs a `ref`

we could also provide a pure `*` interface that does this wrapping.

Reply:
As we introduce more user-friendly ways of passing raw data here, you absolutely have my blessing if you want to change the interface of ```loadChunk``` and ```loadChunk```.

Reply:
I will probably just add two overloads for raw pointers then and will do the wrapping therein.

Reply:
Just so the intention behind the ```shared_ptr``` does not get lost in the ether:
It is supposed to signal that the data behind it must not be changed until the reference count is back to the initial value (i.e. the backend does not hold a reference any more):

```cpp
std::shared_ptr< double > data = /* ... */
Offset offset = /* ... */
Extent extent = /* ... */
long initial =  data.use_count();

auto record_component = /* ... */
record_component.storeChunk(offset, extent, global_data);
std::future< void > future = s.flush();

while( data.use_count() > initial )
{
    /* wörk that DOES NOT modify *data */
}
/* no more such wörk */
future.wait();
/* wörk that DOES modify *data */
```

Reply:
Can you elaborate why it's a `shared` ptr for `stores` but a `unique` ptr for loads? (I understand the in-out meaning of `unique_ptr<T> &` and wonder why we use it only in load)

Reply:
Legacy code and lazyness to refactor it, in all honesty:
https://github.com/openPMD/openPMD-api/blob/62b854a0714e4f6516bc5515f0a4382e673a6b56/include/openPMD/RecordComponent.hpp#L162-L164
The previous interface induced a ```flush()``` and *returned* the pointer, eliminating the need for shared ownership. As the ```unique_ptr``` is more lightweight, it was chosen for that purpose. *This might not have been premature optimization, but now is overly specific.

Reply:
So does it make sense to put both `unique_ptr<T>&` or both `shared_ptr<T>`?

Reply:
[Semantic obviousness](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/) for the user would dictate a ```shared_ptr<T>```. Note that this is also used internally when writing datasets (to use reference-counting):
https://github.com/openPMD/openPMD-api/blob/62b854a0714e4f6516bc5515f0a4382e673a6b56/include/openPMD/IO/IOTask.hpp#L162-L169
**NOTE - This is what the behaviour should be, but currently is not** (see below for explaination).
The downside with ```unique_ptr<T>&``` is that the original ```unique_ptr<T>``` might go out of scope for the user, destroy and deallocate what it points to. The reference would be dangling, and we cannot guarantee the IOTask has been completed by that point:
```cpp
auto record_component = /* ... */
{
    std::unique_ptr< double > data;
    Offset offset = /* ... */
    Extent extent = /* ... */
    record_component.loadChunk(offset, extent, data);
}
s.flush();
```
**NOTE** For some reason I implemented ```loadChunk``` to always [```.flush()```](https://github.com/ax3l/openPMD-api/blob/069a9e2280c3e501007b5b164c27f5dcc375f29f/include/openPMD/RecordComponent.hpp#L195), but ```storeChunk``` to defer flushes with the help of a [queue](https://github.com/ax3l/openPMD-api/blob/069a9e2280c3e501007b5b164c27f5dcc375f29f/include/openPMD/RecordComponent.hpp#L277). Ideally, we would put both into the same queue.

Reply:
Nice, I agree with your points and the ownership arguments in the reference.
I think all in all it makes sense then to make `loadChunk` a `shared_ptr` too.

Reply:
Should we then overload both methods with a "lazy"/"unaware" raw-pointer overload or just expose the `shareRaw` to the user?

Reply:
What are use cases you're thinking of for raw pointers?
Even with such applications, I would rather keep the contract (i.e. shared ownership) in the interface (i.e. offer ```loadChunk```/```storeChunk``` with ```shared_ptr``` only) and offer a helper for raw pointers:
```cpp
template< typename T >
std::shared_ptr< T >
storeRaw( T* ptr )
{
    return std::shared_ptr< T >( ptr, [](T*){} );
}
```

Reply:
This mechanism does not offer the safety-guarantees I referred to in my previous comment, btw:
```cpp
auto record_component = /* ... */
{
    std::vector< double > data = /* ... */;
    Offset offset = /* ... */
    Extent extent{data.size()};
    record_component.storeChunk(offset, extent, storeRaw(data));
}
//pointer behind `data` does not exist any more, but flush might not have happened
s.flush();
```

Reply:
> What are use cases you're thinking of for raw pointers?

Mainly contiguous containers (STL and third-party libs), that don't offer a shared view or unique-ptr into their data.

Reply:
While refactoring the `loadChunk` interface to `shared_ptr` I realized the following: For proper support of the "we allocate for you" API I would need to pass as `shared_ptr< T[] >&` which I find a little odd. Also, the implementation allocated immediately which does not offer too much convenience.

What do you think about if I:
- remove the allocation policy argument
- use a signature (1) that returns a `shared_ptr<>` (we allocate)
- use a signature (2) that accepts a pre-allocated `shared_ptr<>` (user allocates)

Reply:
Overall I'll have to agree. Didn't think far enough to realize that you can't allocate the ```shared_ptr<T>``` and have it visible to the caller. The proposal with the two signatures also makes sense. Do you have any catchy names for semantic differentiation (i.e. ```loadChunkAllocated```)? 

> use a signature (1) that returns a shared_ptr<T> (we allocate)

Good thing I was lazy enough to leave that code in there.

Reply:
I don't thinkg we will need different names, the parameters are different :-)

Reply:
Btw, I think [this example](https://github.com/openPMD/openPMD-api/pull/204#discussion_r190829625) you posted would make a great page in the manual! :)

Reply:
> ```shared_ptr< T[] >``` machinery was kind of forgotten in C++11

How this is only coming in C++17 is beyond me...